README.md
# FallingSky
Demo Java JFrame Game


ChangeLog.txt
V 0.0.1

Basic game with a home screen and simple gameplay. 
Base obstacle and HoleInTheWall implemented.
Has score in the corner and the blocks move faster as time goes on capped at 5.

src/Player.java
package fallingSky;

import java.util.*;
import java.awt.*;
import java.awt.event.*;

public class Player implements KeyListener {

	ArrayList<Integer> keys;

	private double x;

	private double y;

	private double yVel;

	private double xVel;

	private double maxSpeed;

	public Player () {

		keys = new ArrayList<Integer>();

		x = 520;

		y = 200;

		xVel = 0;

		yVel = 0;

		maxSpeed = 10;

	}

	public void keyPressed (KeyEvent e) {

		if (!keys.contains(e.getKeyCode())) {

			keys.add((Integer) e.getKeyCode());

		}

	}

	public void keyTyped (KeyEvent e) {


	}

	public void keyReleased (KeyEvent e) {

		keys.remove((Integer) e.getKeyCode());

	}

	public ArrayList<Integer> getKeys() {

		return keys;

	}

	public double getX () {

		return x;

	} 

	
	public double getY () {

		return y;

	}

	public double getXVel () {

		return xVel;

	}

	public double getYVel () {

		return yVel;

	}

	public void setXVel (double value) {

		xVel = value;

	}

	public void setYVel (double value) {

		yVel = value;

	}

	public void move () {

		if (Math.abs(xVel) > 0.1) {

			x += xVel;

		}

		y += yVel;

	}

	public void stop () {

		xVel = 0;

		yVel = 0;

	}

	public void changeVelocity () {

		if (keys.contains(KeyEvent.VK_A) && x > 0) {


			if (xVel > - (maxSpeed)) {

				xVel -= 0.5;

			} else {

				xVel = - (maxSpeed);

			}

		} else if (keys.contains(KeyEvent.VK_D) && x < 1040) {


			if (xVel < maxSpeed) {

				xVel += 0.5;

			} else {

				xVel = maxSpeed;

			}


		} else {
	
			if (xVel > 0) {

				xVel -= 0.1;

			} else if (xVel < 0) {

				xVel += 0.1;

			} else {

				stop();

			}

		}

		if (x < 0) {

			x = 0;

			stop();

		} else if (x > 1040) {

			x = 1040;

			stop();

		}
	}

	public boolean isTouchingObstacle (Obstacle o) {

		return o.isTouchingPlayer(this);

	}

}

src/Obstacle.java
package fallingSky;

import java.awt.*;

public class Obstacle {

	private double x;

	private double y;

	private int width;
	
	private int height;

	private double xVel;

	private double yVel;

	public Obstacle () {

		y = 720;

		x = (Math.random() * 780);
		
		width = 300;

		height = 100;

		xVel = 0;

		yVel = -4;

	}

	public Obstacle (double x, double y, int width, int height) {

		this.x = x;

		this.y = y;

		this.width = width;

		this.height = height;

		xVel = 0;

		yVel = -4;

	}


	public void move () {

		if (xVel < 0 && x < 0) {

			xVel = -(xVel); 			

		} else if (xVel > 0 && x > 1080 - width) {

			xVel = -(xVel);

		}

		x += xVel;

		y += yVel;

	}

	public double getX () {

		return x;

	}

	public double getY () {

		return y;

	}

	public double getXVel () {

		return xVel;

	}

	public double getYVel () {

		return yVel;

	}

	public int getWidth () {

		return width;

	}

	public int getHeight () {

		return height;

	}

	public void setXVel (double value) {

		xVel = value;

	}

	public void setYVel (double value) {

		yVel = value;

	}

	public void paintObstacle (Graphics g) {


		g.fillRect((int) getX(), (int) getY(), getWidth(), getHeight());


	}

	public boolean isTouchingPlayer (Player p) {

		if (p.getX() + 40 < getX() || p.getX() > getX() + getWidth()) {

			return false;

		}

		if (p.getY() + 40 < getY() || p.getY() > getY() + getHeight()) {

			return false;

		}

		return true;

	}

}

src/Renderer.java
package fallingSky;

import java.awt.event.*;
import javax.swing.*;
import java.awt.*;

public class Renderer implements ActionListener, MouseListener {

	private Timer t;

	private Window w;

	private int screen;

	private Player p;

	private Obstacle obstacle;

	private int score;

	public Renderer() {

		w = new Window();

		screen = 0;

		w.addMouseListener(this);

		t = new Timer(10, this);

		t.start();

	}

	public void actionPerformed (ActionEvent e) {

		if (screen == 1) {

			p.changeVelocity();
			
			p.move();

			if (obstacle == null || obstacle.getY() < 0 - obstacle.getHeight()) {

				int spawnType = (int) (Math.random() * 2);

				double speed = -2 - (score / 100) / 10.0;

				if (score >= 2000) {

					speed = -4;

				}		

				if (spawnType == 0) {

					obstacle = new Obstacle();

					obstacle.setXVel((Math.random() * 10) - 5);

					obstacle.setYVel(speed);
					

				} else if (spawnType == 1) {
			
					obstacle = new HoleInTheWall();
		
					obstacle.setYVel(speed);

				}

				w.setObstacle(obstacle);

			}

			if (obstacle != null) {

				obstacle.move();

				if (p.isTouchingObstacle(obstacle)) {

					screen = 0;
					w.setScreen(0);
					p = null;
					obstacle = null;

				}

			}

			score++;

			w.setScore(score);

		}

		w.repaint();

	}

	public void mouseClicked (MouseEvent e) {

		if (screen == 0) {

			if ((e.getX() > 440 && e.getX() < 640) && (e.getY() > 400 && e.getY() < 480)) {

				screen = 1;

				p = new Player();

				w.setPlayer(p);

				w.setScreen(1);

				score = 0;

				w.setScore(score);

			}

		}

	}

	public void mouseEntered (MouseEvent e) {



	}

	public void mouseExited (MouseEvent e) {



	}

	public void mousePressed (MouseEvent e) {



	}

	public void mouseReleased (MouseEvent e) {



	}

}

src/HoleInTheWall.java
package fallingSky;

import java.awt.*;

public class HoleInTheWall extends Obstacle {

	private Obstacle rightO;

	public HoleInTheWall () {

		super(0, 720, (int) (Math.random() * 780) + 100, 100);

		int leftWidth = this.getWidth();

		rightO = new Obstacle(leftWidth + 100, 720, 1080 - leftWidth - 100, 100);

	}

	public void move() {

		super.move();

		rightO.move();

	}

	public Obstacle getRightSide () {

		return rightO;

	}

	public void setYVel (double value) {

		super.setYVel(value);

		rightO.setYVel(value);

	}

	public void paintObstacle (Graphics g) {

		super.paintObstacle(g);

		rightO.paintObstacle(g);

	}

	public boolean isTouchingPlayer (Player p) {

		if (super.isTouchingPlayer(p) == true) {

			return true;

		}

		if (rightO.isTouchingPlayer(p) == true) {

			return true;

		}

		return false;
		

	}


}

src/Window.java
package fallingSky;

import javax.swing.*;
import java.awt.*;

public class Window extends JPanel {

	private JFrame window;

	private int screen;

	private int titleX;
	
	private int titleY;

	private Player p;

	private Obstacle obstacle;

	private int score;

	public Window() {

		window = new JFrame();

		window.setVisible(true);

		window.setResizable(false);

		window.setFocusable(true);

		window.setBackground(new Color(120, 200, 255));

		window.getContentPane().setBackground(new Color(120, 220, 255));

		window.setTitle("Falling Sky");

		window.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);

		window.setSize(1086, 755);

		window.add(this);

		screen = 0;

		titleX = 220;

		titleY = -20;

	}

	public void setScreen (int screenNumber) {

		screen = screenNumber;

	}

	public void setScore (int score) {

		this.score = score;

	}

	public void paintComponent (Graphics g) {

		if (screen == 0) {

			paintTitleScreen(g);

		} else if (screen == 1) {

			paintGame(g);

		}


	}

	public void setPlayer (Player p) {

		this.p = p;

		this.addKeyListener(p);

		window.addKeyListener(p);

	}

	public void setObstacle (Obstacle o) {

		obstacle = o;

	}

	private void paintTitleScreen(Graphics g) {

		g.setColor(new Color(120, 200, 255));

		g.fillRect(0, 0, 1080, 720);

		g.setColor(new Color(150, 150, 150));
		
		g.fillRect(440, 400, 200, 80);

		g.setColor(Color.WHITE);

		g.setFont(new Font(Font.SANS_SERIF, Font.BOLD, 60));

		g.drawString("Start", 470, 460);

		g.setFont(new Font(Font.SANS_SERIF, Font.BOLD, 40));

		g.drawString("A HalfCookedDumpling Creation", 235, 680);

		g.setFont(new Font(Font.SANS_SERIF, Font.BOLD, 120));

		g.drawString("Falling Sky", titleX, titleY);

		if (titleY < 280) {

			titleY += 5;

		}

	}

	private void paintGame (Graphics g) {

		g.setColor(new Color(120, 200, 255));

		g.fillRect(0, 0, 1080, 720);

		g.setColor(new Color(150, 150, 150));

		g.fillRect((int) p.getX(), (int) p.getY(), 40, 40);

		if (obstacle != null) {

			g.setColor(Color.RED);

			obstacle.paintObstacle(g);
		}

		g.setColor(Color.WHITE);

		g.setFont(new Font(Font.SANS_SERIF, Font.BOLD, 20));

		g.drawString("Score: " + score, 10, 30);

	}


}